server/src/main/java/org/activityinfo/ui/client/component/form/field/attachment/Uploader.java

```diff
@@ -0,0 +1,146 @@
+import java.util.Map;
+
+/**
+ * @author yuriyz on 11/16/2015.
+ */
+public class Uploader {
+
+    public interface UploadCallback {
+        void onFailure(@Nullable Throwable exception);
+
+        void upload();
+    }
+
+    private final FileUpload fileUpload;
+    private final FormPanel formPanel;
+    private final VerticalPanel hiddenFieldsContainer;
+    private final UploadCallback uploadCallback;
+
+    private final Holder<Attachment> attachment;
+
+    public Uploader(FormPanel formPanel, FileUpload fileUpload, Holder<Attachment> attachment,
+                    VerticalPanel hiddenFieldsContainer, UploadCallback uploadCallback) {
+        this.formPanel = formPanel;
+        this.fileUpload = fileUpload;
+        this.attachment = attachment;
+        this.hiddenFieldsContainer = hiddenFieldsContainer;
+        this.uploadCallback = uploadCallback;
+    }
+
+    public void requestUploadUrl() {
+        try {
+            RequestBuilder requestBuilder = new RequestBuilder(RequestBuilder.POST, URL.encode(createUploadUrl()));
+            requestBuilder.sendRequest(null, new RequestCallback() {
+                @Override
+                public void onResponseReceived(Request request, Response response) {
+
+                    Resource resource = Resources.fromJson(response.getText());
+                    UploadCredentials uploadCredentials = UploadCredentials.fromRecord(resource);
+
+                    removeHiddenFieldsFromForm();
+
+                    Map<String, String> formFields = uploadCredentials.getFormFields();
+                    for (Map.Entry<String, String> field : formFields.entrySet()) {
+                        hiddenFieldsContainer.add(new Hidden(field.getKey(), field.getValue()));
+                    }
+
+                    formPanel.setAction(uploadCredentials.getUrl());
+                    formPanel.setMethod(uploadCredentials.getMethod());
+                    uploadCallback.upload();
+                }
+
+                @Override
+                public void onError(Request request, Throwable exception) {
+                    Log.error("Failed to send request", exception);
+                    uploadCallback.onFailure(exception);
+                }
+            });
+        } catch (Exception e) {
+            Log.error("Failed to send request", e);
+            uploadCallback.onFailure(e);
+        }
+    }
+
+    private String createUploadUrl() {
+        try {
+            String blobId = ResourceId.generateId().asString();
+            String fileName = fileName();
+            String mimeType = MimeTypeUtil.mimeTypeFromFileName(fileName);
+
+            attachment.get().setMimeType(mimeType);
+            attachment.get().setFilename(fileName);
+            attachment.get().setBlobId(blobId);
+            return "/service/blob/credentials/" + blobId;
+        } catch (Exception e) {
+            Log.error(e.getMessage(), e);
+            Window.alert(I18N.CONSTANTS.failedToCreateUploadUrl());
```

Comment:
Yes, we can just report that upload failed. Background is that user may upload file which is not supported in by our MimeTypeUtil and therefore mimtype indentification fails. Like .png -> image/png. (it is very rare case because we have mapping for more than 200 file extension.) What retry path would you propose in case we fail to identify mime type ?